base/sparse/sparsematrix.jl

```diff
@@ -2597,10 +2600,12 @@ function setindex!{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractMatrix{Bool})
 
                 if (mode > 1) && (nadd == 0) && (ndel == 0)
                     # copy storage to take changes
-                    colptrB = copy(colptrA)
+                    colptrA = copy(colptrB)
                     memreq = (x == 0) ? 0 : n
-                    rowvalB = Array(Ti, length(rowvalA)+memreq); copy!(rowvalB, 1, rowvalA, 1, r1-1)
-                    nzvalB = Array(Tv, length(nzvalA)+memreq); copy!(nzvalB, 1, nzvalA, 1, r1-1)
+                    rowvalA = copy(rowvalB)
+                    nzvalA = copy(nzvalB)
+                    resize!(rowvalB, length(rowvalA)+memreq)
+                    resize!(nzvalB, length(rowvalA)+memreq)
```

Comment:
> You basically want to swap xxxA and xxxB?

Eventually something along those lines might be nice, yes. But for now please ignore this comment (see below.)

> I agree, these functions are indeed difficult to reason about. They try to reuse the storage in the input matrix but if they find out that they need temporary data they allocate copies of the matrix fields and rebind the variables. It makes it difficult to keep track on what data each variable points to. I didn't want to dig down too much into these functions so I mostly mechanically made them set the new values inplace the matrix fields instead of the newly allocated vectors.

Cheers, I would have done the same and for the same reasons. I like how you've minimized modifications in this PR. Better to defer my suggestion above to a separate PR to retain that approach and see this merged efficiently.